# Day 3: Wednesday August 22

### General Topics

- Introduction to Machine Learning and Data Mining
  - Machine Learning basic concepts
  - Taxonomy of ML algorithms
    - Supervised classification,
    - Unsupervised classification,
    - Semi-supervised classification
  - Clustering validation and cross validation 

- Applications of ML in Bioinformatics
  - Examples of different ML/DM techniques that can be applied to different NGS data analysis pipelines
  - How to chose the right ML technique?

  
- Deep learning quick overview

- Exploring some examples with R ML packages

## Schedule

- _**09:00 - 10:00**_ Introduction to  DM and ML, Machine Learning basic concepts

- _**10:00 - 11:00**_ Taxonomy of ML and examples of algorithms
    - Supervised classification,
    - Unsupervised classification,
    - Semi-supervised classification

- _**11:00 - 11:30**_ _**Coffee break**_

- _**11:30 - 12:30**_ Applications of ML in Bioinformatics
    - Examples of different ML/DM techniques that can be applied to different NGS data analysis pipelines
    - How to chose the right ML technique?
    - Deep learning overview

- _**12:30 - 14:00**_ _**Lunch break**_

- _**14:00 - 15:00**_ We will start practicing using the built-in R [data set iris](https://archive.ics.uci.edu/ml/datasets/iris)

- _**15:00 - 16:00**_ RNASeq analysis using clustering in R <br/>
[Example we will be using during the session](http://www.2-bitbio.com/2017/04/clustering-rnaseq-data-making-heatmaps.html)<br/>



- _**16:00 - 16:15**_ _**Coffee break**_

- _**16:15 - 18:00**_ RNASeq analysis in R to be continued

RNASeq analysis in R <br/>
[Differential expression analysis using R](https://combine-australia.github.io/RNAseq-R/06-rnaseq-day1.html) <br/>
Download Mouse mammary RNA-Seq data (counts) data files from [this public repo](https://figshare.com/s/1d788fd384d33e913a2a)<br/>

## Let's start doing some clustering on the IRIS dataset

```
iris #to check the dataset
```
Preprocessing the data: exclude column "Species" at position 5 and Standardize

```
F_iris <- iris[,-5] # Remove column 5
F_iris <- scale(F_iris) # Standardize: Scale is a generic function whose default method centers and/or scales the columns of a numeric matrix.
```

Let's cluster the data using function eclust() (enhanced clustering, in factoextra). You will need to install factoextra package first.<br/> 

Install Bioconductor packages. More details [here](https://www.bioconductor.org/install/) <br/>

```
source("https://bioconductor.org/biocLite.R")
biocLite()
```
Install factoextra
```
biocLite("factoextra")
```
Load factoextra
```
library(factoextra)
```

For more info about the eclust() function, check [this](https://www.rdocumentation.org/packages/factoextra/versions/1.0.5/topics/eclust)
or type
```
help(eclust)
eclust(x, FUNcluster = "kmeans", hc_metric = "euclidean", ...) 
```
Now let's try to cluster the data with Kmeans and then use hierarchical clustering 

```
eclust(F_iris, FUNcluster = "kmeans", hc_metric = "euclidean") #number of clusters not specified, will be detected automatically
# K-means clustering giving number of clusters= 3
km.res <- eclust(df, "kmeans", k = 3, nstart = 25, graph = FALSE)
# Visualize k-means clusters
fviz_cluster(km.res, geom = "point", ellipse.type = "norm", palette = "jco", ggtheme = theme_minimal())
```

```
hc.res <- eclust(F_iris, "hclust", k = 3, hc_metric = "euclidean", hc_method = "ward.D2", graph = FALSE)
# Visualize dendrograms
fviz_dend(hc.res, show_labels = FALSE, palette = "jco", as.ggplot = TRUE)
 ```
Let's check the clusters quality, remember we talked this morning about clustering internal validation and the use of coefficients and indexes to measure clusters separability and intra clusters homogenity. We will compute the silhouette coefficient for the resulting K-means clustering

```
fviz_silhouette(km.res, palette = "jco", ggtheme = theme_classic())
```
A value close to 1 indicates good quality clustering (element assigned to the right cluster)<br/>
A value close to -1 indicates poor quality clustering (element would better fit in another cluster)<br/>

Now try to vary the number of clusters, run Kmeans and compute the the silhouette coefficient. What did you notice? <br/>

Many other coefficients could be computed to assess the clustering quality, some are available using the cluster.stats() function of the fpc package. You might need to install the fpc package first.
```
biocLite("fpc")
library(fpc)
kmIris_stats <- cluster.stats(dist(F_iris),  km.res$cluster)
kmIris_stats$dunn # if you wanna see only the dunn index value
```

## Some interesting readings: application of ML in Bioinformatics

[A machine learning model to determine the accuracy of variant calls in capture-based next generation sequencing](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-018-4659-0) <br/>
[Next-Generation Machine Learning for Biological Networks](https://www.ncbi.nlm.nih.gov/pubmed/29887378) <br/>
[Machine learning for metagenomics: methods and tools](https://arxiv.org/pdf/1510.06621.pdf) <br/>

## Some useful R/Bioconductor packages to handle Bioinformatics files and use ML algorithms

[vcfR, a package to read, manipulate and analyze VCF files](https://knausb.github.io/vcfR_documentation/index.html)<br/>
[Practical Guide To Cluster Analysis in R](https://goo.gl/13EFCZ)<br/>

### Lead Instructor
- [Amel Ghouila](https://github.com/amelgh) | [@AmelGhouila](https://twitter.com/AmelGhouila) | [email](mailto:amel.ghouila@gmail.com)

### Co-Instructor(s)
- [Fotis Psomopoulos](https://fpsom.github.io/) | [@fopsom](https://twitter.com/fopsom) | [email](mailto:fopsom@gmail.com)

### Helper(s)
- Maria Tsagiopoulou | [email](mariatsayo@gmail.com)
- Ola Karrar | [email](ozkarrar2@gmail.com)
